File: apps/web/ios/App/App/HeadingBridge.swift
import Foundation
import CoreLocation
import UIKit
import Capacitor

/**
 * HeadingBridge — the native compass feed for the qiblah compass.
 *
 * The web layer cannot get a heading inside a WKWebView: `webkitCompassHeading`
 * is a Safari-only extension and is absent in an app's web view, and
 * DeviceOrientationEvent's `alpha` is not north-referenced there, so it can't be
 * turned into a bearing. (@capacitor/motion is only a web shim over those same
 * events, so it inherits the problem.) CoreLocation is the real source, which is
 * all this plugin is: CLLocationManager.startUpdatingHeading() streamed to JS.
 *
 * JS side — src/lib/mobile/heading.ts / components/QiblahSection.tsx:
 *   const h = await HeadingBridge.addListener("heading", (e) => { … })
 *   await HeadingBridge.start()
 *   …
 *   h.remove(); await HeadingBridge.stop()
 *
 * Event payload:
 *   { trueHeading: Double, magneticHeading: Double, accuracy: Double }
 *
 * `trueHeading` ALREADY includes the local magnetic declination — the JS side
 * must NOT apply its own correction on top of it. It is NEGATIVE when
 * CoreLocation has no location fix to derive declination from; the caller then
 * falls back to `magneticHeading` and corrects that itself. `accuracy` is
 * CLHeading.headingAccuracy: degrees of maximum error, negative when the reading
 * is invalid and the magnetometer needs calibrating.
 *
 * Location permission is already granted for prayer times, so this adds no new
 * consent prompt (heading updates need no authorization of their own; the
 * authorization only decides whether a TRUE heading is available).
 */
@objc(HeadingBridge)
public class HeadingBridge: CAPPlugin, CAPBridgedPlugin, CLLocationManagerDelegate {
    public let identifier = "HeadingBridge"
    public let jsName = "HeadingBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise)
    ]

    /// Created lazily on the main thread — CLLocationManager must be used from a
    /// thread with an active run loop, and Capacitor calls arrive on a background
    /// queue.
    private var manager: CLLocationManager?
    /// Gates event emission so a delivery already in flight when stop() lands
    /// can't reach a compass the web layer has already torn down.
    private var running = false
    private var observingOrientation = false

    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve(["available": CLLocationManager.headingAvailable()])
    }

    @objc func start(_ call: CAPPluginCall) {
        // iPads and the simulator have no magnetometer; reject rather than
        // resolve into a stream that never emits, so the UI can say so honestly.
        guard CLLocationManager.headingAvailable() else {
            call.reject("This device has no compass (magnetometer) available")
            return
        }
        DispatchQueue.main.async {
            let manager = self.manager ?? CLLocationManager()
            self.manager = manager
            manager.delegate = self
            // 1° — the dial animates between readings anyway, and anything finer
            // is noise the user can't act on.
            manager.headingFilter = 1
            // iOS reports the heading of whichever device edge headingOrientation
            // names, so a rotated phone (the app allows landscape) reads 90° off
            // unless this tracks the real orientation.
            self.syncHeadingOrientation()
            self.observeOrientation()
            self.running = true
            manager.startUpdatingHeading()
            call.resolve()
        }
    }

    @objc func stop(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.running = false
            self.manager?.stopUpdatingHeading()
            NotificationCenter.default.removeObserver(
                self, name: UIDevice.orientationDidChangeNotification, object: nil)
            if self.observingOrientation {
                UIDevice.current.endGeneratingDeviceOrientationNotifications()
            }
            self.observingOrientation = false
            call.resolve()
        }
    }

    private func observeOrientation() {
        guard !observingOrientation else { return }
        observingOrientation = true
        // UIKit contract: UIDevice.orientation returns .unknown and the
        // orientationDidChange notification NEVER fires unless generation is
        // explicitly begun. Balanced by endGenerating in stop().
        UIDevice.current.beginGeneratingDeviceOrientationNotifications()
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(orientationDidChange),
            name: UIDevice.orientationDidChangeNotification,
            object: nil)
    }

    @objc private func orientationDidChange() {
        syncHeadingOrientation()
    }

    /// CLDeviceOrientation shares UIDeviceOrientation's raw values. Face-up /
    /// face-down / unknown aren't interface orientations, so those are skipped and
    /// the last real one is kept — a phone held flat (exactly how you use a
    /// compass) reports faceUp constantly.
    private func syncHeadingOrientation() {
        let deviceOrientation = UIDevice.current.orientation
        guard deviceOrientation.isValidInterfaceOrientation,
              let orientation = CLDeviceOrientation(rawValue: Int32(deviceOrientation.rawValue))
        else { return }
        manager?.headingOrientation = orientation
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        guard running else { return }
        notifyListeners("heading", data: [
            "trueHeading": newHeading.trueHeading,
            "magneticHeading": newHeading.magneticHeading,
            "accuracy": newHeading.headingAccuracy
        ])
    }

    /// Let iOS put up its own figure-8 calibration sheet when it decides the
    /// magnetometer needs it — the single biggest cause of a wrong qiblah.
    public func locationManagerShouldDisplayHeadingCalibration(_ manager: CLLocationManager) -> Bool {
        return true
    }
}

File: apps/web/android/app/src/main/java/com/hiddenhiqmah/app/HeadingBridge.java
package com.hiddenhiqmah.app;

import android.content.Context;
import android.hardware.GeomagneticField;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.location.Location;
import android.location.LocationManager;
import android.view.Surface;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Compass heading — the Android twin of ios/App/App/HeadingBridge.swift.
 *
 * Same JS contract (src/lib/mobile/heading.ts): start/stop/isAvailable plus a
 * "heading" event carrying trueHeading, magneticHeading and accuracy. The web
 * path can't serve as a fallback here for the same reason it can't on iOS —
 * DeviceOrientationEvent.alpha is not north-referenced in a WebView.
 *
 * ⚠️ THE DIFFERENCE THAT MATTERS: iOS hands you `trueHeading` with magnetic
 * declination ALREADY APPLIED. Android does not. The rotation vector is
 * referenced to MAGNETIC north, and true north is magnetic + declination, which
 * we have to compute ourselves from GeomagneticField at the user's position.
 * Get this wrong and the qiblah is off by the local declination — up to ~20° in
 * parts of North America, which is the difference between facing Makkah and
 * facing the next city over.
 *
 * Declination is applied EXACTLY ONCE, here, so it matches the iOS contract
 * where trueHeading arrives ready to use. QiblahSection has a single dial entry
 * point (applyTrueHeading) precisely so it can never be applied twice.
 */
@CapacitorPlugin(name = "HeadingBridge")
public class HeadingBridge extends Plugin implements SensorEventListener {

    private SensorManager sensorManager;
    private Sensor rotationVector;
    private boolean running = false;

    /** Degrees to add to a magnetic heading to get a true one. 0 until we have a fix. */
    private float declination = 0f;
    private boolean haveDeclination = false;

    private final float[] rotationMatrix = new float[9];
    private final float[] orientation = new float[3];

    /**
     * The rotation vector updates far faster than a compass dial needs, and every
     * event crosses the JS bridge. ~20/s is smooth to the eye and an order of
     * magnitude cheaper than the raw stream.
     */
    private static final long MIN_INTERVAL_MS = 50;
    private long lastEmit = 0L;

    @Override
    public void load() {
        sensorManager = (SensorManager) getContext().getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager != null) {
            // GEOMAGNETIC_ROTATION_VECTOR uses the magnetometer without the
            // gyroscope: slightly less smooth, but it is the one that stays
            // north-referenced and it costs far less battery. Fall back to the
            // full rotation vector on devices that lack it.
            rotationVector = sensorManager.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR);
            if (rotationVector == null) {
                rotationVector = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
            }
        }
    }

    @PluginMethod
    public void isAvailable(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("available", rotationVector != null);
        call.resolve(ret);
    }

    @PluginMethod
    public void start(PluginCall call) {
        if (rotationVector == null) {
            call.reject("No compass on this device");
            return;
        }
        if (!running) {
            refreshDeclination();
            sensorManager.registerListener(this, rotationVector, SensorManager.SENSOR_DELAY_GAME);
            running = true;
        }
        call.resolve();
    }

    @PluginMethod
    public void stop(PluginCall call) {
        stopUpdates();
        call.resolve();
    }

    private void stopUpdates() {
        if (running && sensorManager != null) {
            sensorManager.unregisterListener(this);
            running = false;
        }
    }

    /** The magnetometer keeps costing battery if the activity goes away mid-session. */
    @Override
    protected void handleOnPause() {
        stopUpdates();
    }

    @Override
    protected void handleOnDestroy() {
        stopUpdates();
    }

    /**
     * Declination for the user's current position. Uses the last known fix only —
     * this never requests location itself, because the app has already asked for
     * it for prayer times and a compass should not trigger a second prompt. With
     * no fix we leave declination at 0 and report trueHeading as -1, which the JS
     * contract defines as "not available", so the caller applies its own
     * correction (QiblahSection carries a WMM model for exactly this).
     */
    private void refreshDeclination() {
        try {
            LocationManager lm = (LocationManager) getContext().getSystemService(Context.LOCATION_SERVICE);
            if (lm == null) return;
            Location best = null;
            for (String provider : lm.getProviders(true)) {
                Location loc = lm.getLastKnownLocation(provider);
                if (loc == null) continue;
                if (best == null || loc.getTime() > best.getTime()) best = loc;
            }
            if (best == null) return;
            GeomagneticField field = new GeomagneticField(
                    (float) best.getLatitude(),
                    (float) best.getLongitude(),
                    (float) best.getAltitude(),
                    System.currentTimeMillis());
            declination = field.getDeclination();
            haveDeclination = true;
        } catch (SecurityException e) {
            // Permission revoked mid-session: magnetic heading still works.
            haveDeclination = false;
        }
    }

    /**
     * Screen rotation changes which axis points "up", and the sensor frame does
     * not follow it. iOS handles this via headingOrientation; here we remap the
     * rotation matrix ourselves, or a landscape device reads 90° off.
     */
    private int displayRotation() {
        try {
            return getActivity().getWindowManager().getDefaultDisplay().getRotation();
        } catch (Exception e) {
            return Surface.ROTATION_0;
        }
    }

    /** Drop the filter state so a stale bearing cannot bleed into a new session. */
    private void resetSmoothing() {
        haveSmoothed = false;
        haveLastEmitted = false;
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor != rotationVector) return;

        long now = System.currentTimeMillis();
        if (now - lastEmit < MIN_INTERVAL_MS) return;
        lastEmit = now;

        SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values);

        int axisX = SensorManager.AXIS_X;
        int axisY = SensorManager.AXIS_Y;
        switch (displayRotation()) {
            case Surface.ROTATION_90:
                axisX = SensorManager.AXIS_Y;
                axisY = SensorManager.AXIS_MINUS_X;
                break;
            case Surface.ROTATION_180:
                axisX = SensorManager.AXIS_MINUS_X;
                axisY = SensorManager.AXIS_MINUS_Y;
                break;
            case Surface.ROTATION_270:
                axisX = SensorManager.AXIS_MINUS_Y;
                axisY = SensorManager.AXIS_X;
                break;
            default:
                break;
        }
        float[] remapped = new float[9];
        SensorManager.remapCoordinateSystem(rotationMatrix, axisX, axisY, remapped);
        SensorManager.getOrientation(remapped, orientation);

        // azimuth is radians from magnetic north, -pi..pi
        double magnetic = Math.toDegrees(orientation[0]);
        magnetic = (magnetic + 360.0) % 360.0;

        magnetic = smooth(magnetic);

        // Suppress sub-degree churn. Below this the needle is only rendering
        // sensor noise, and every emission crosses the JS bridge and triggers a
        // re-render — so the jitter costs battery as well as looking unsteady.
        if (haveLastEmitted && angularDelta(magnetic, lastEmittedMagnetic) < MIN_DELTA_DEG) {
            return;
        }
        haveLastEmitted = true;
        lastEmittedMagnetic = magnetic;

        // Declination applied exactly once — see the class note.
        double trueHeading = haveDeclination ? (magnetic + declination + 360.0) % 360.0 : -1.0;

        JSObject ret = new JSObject();
        ret.put("trueHeading", trueHeading);
        ret.put("magneticHeading", magnetic);
        ret.put("accuracy", accuracyDegrees());
        notifyListeners("heading", ret);
    }

    /**
     * Low-pass the heading.
     *
     * The raw rotation vector is noisy enough that the needle visibly shakes
     * while you hold the phone still — reported from the device. iOS never
     * needed this because Core Location hands back an already-smoothed
     * trueHeading; Android gives the raw thing.
     *
     * ⚠️ Averaged as a UNIT VECTOR, never as degrees. A naive average of 359°
     * and 1° is 180° — the needle would slam to the opposite direction every
     * time the user faced north, which is precisely where a qibla compass must
     * be most trustworthy. Converting to (cos, sin), smoothing each component
     * and taking atan2 back is the circular mean, which has no such seam.
     *
     * ⭐ ADAPTIVE, because a fixed alpha cannot win. A single smoothing constant
     * trades one complaint for the other: high enough to track a real turn and
     * the needle still shakes at rest; low enough to be steady at rest and the
     * needle lags behind the phone. 0.18 was still visibly finicky on device.
     *
     * So the alpha follows the SIZE of the change. A small delta is almost
     * certainly noise — a phone held still does not rotate two degrees — and gets
     * heavy smoothing. A large delta is the user actually turning, and gets very
     * little, so the needle keeps up. The result is steady in the hand and still
     * responsive, which is what a qibla compass has to be: people rotate slowly,
     * hunting the last few degrees, exactly where noise is worst.
     */
    private static final double ALPHA_STILL = 0.045; // noise: smooth hard
    private static final double ALPHA_TURN = 0.55; // real motion: track it
    private static final double TURN_DEG = 8.0; // above this, treat as a turn
    private static final double MIN_DELTA_DEG = 1.0;
    private double smoothX, smoothY;
    private boolean haveSmoothed = false;
    private double lastEmittedMagnetic;
    private boolean haveLastEmitted = false;

    private double smooth(double degrees) {
        double rad = Math.toRadians(degrees);
        double x = Math.cos(rad), y = Math.sin(rad);
        if (!haveSmoothed) {
            smoothX = x;
            smoothY = y;
            haveSmoothed = true;
            return degrees;
        }
        double current = (Math.toDegrees(Math.atan2(smoothY, smoothX)) + 360.0) % 360.0;
        double delta = angularDelta(degrees, current);

        // Ramp between the two constants rather than switching, so a turn that
        // starts gently does not visibly "click" into a faster mode.
        double t = Math.min(1.0, delta / TURN_DEG);
        double alpha = ALPHA_STILL + (ALPHA_TURN - ALPHA_STILL) * t * t;

        smoothX += alpha * (x - smoothX);
        smoothY += alpha * (y - smoothY);
        return (Math.toDegrees(Math.atan2(smoothY, smoothX)) + 360.0) % 360.0;
    }

    /** Shortest angle between two bearings, 0..180 — wrap-safe. */
    private static double angularDelta(double a, double b) {
        double d = Math.abs(a - b) % 360.0;
        return d > 180.0 ? 360.0 - d : d;
    }

    /**
     * The JS contract wants a maximum error in degrees, NEGATIVE when the reading
     * is invalid and the magnetometer wants calibrating — the same convention as
     * webkitCompassAccuracy. Android reports a coarse enum instead, so map it.
     */
    private double accuracyDegrees() {
        switch (lastAccuracy) {
            case SensorManager.SENSOR_STATUS_ACCURACY_HIGH:
                return 5.0;
            case SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM:
                return 15.0;
            case SensorManager.SENSOR_STATUS_ACCURACY_LOW:
                return 30.0;
            default:
                return -1.0;
        }
    }

    private int lastAccuracy = SensorManager.SENSOR_STATUS_ACCURACY_HIGH;

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        if (sensor == rotationVector) lastAccuracy = accuracy;
    }
}
